### ISSUE TITLE
Anybody found a way to use enums in the iP?

### ISSUE THREAD

**Issue by damithc** [nus-cs2103-AY2021S2/forum#14]
The iP asks you to use Java `enum`s, _if applicable_.

If you have found a way to use them in your iP, feel free to share (i.e., for what did you use it) here for the benefit of others.

**Comment by w2vgd** [comment]
Have not implemented this yet, but I am guessing we can use an enum for the commands (e.g. bye, list, todo, event, deadline, etc.)

**Comment by Cheng20010201** [comment]
Agree with @w2vgd. I used enum for the commands to facilitate printing out a help menu for commands available (just for fun).
However, for the actual functionality side I haven't figured out a very elegant way to integrate enum with commands. Any suggestions?

**Comment by JonahhGohh** [comment]
I used enums for the commands and a switch statement in the main method. For the enum class, I used a [constructor for the enum constants](https://dzone.com/articles/usage-enumeration-enum-java#:~:text=We%20use%20Enum%20for%20a,final%20keywords%20to%20represent%20constants.&text=We%20can%20define%20values%20for,field%20variable%20for%20this%20enum.) to give them a String value. This is so that I can relate the enum commands to the commands inputted by the user.
e.g. `TODO("todo")`
However since the commands in my main were of String type, I converted them to the enum type using a [lookup method](https://stackoverflow.com/questions/1080904/how-can-i-lookup-a-java-enum-from-its-string-value) and only after that pass the commands into a switch statement.

**Comment by chesterhow** [comment]
I thought of using enums for the commands too, but as @JonahhGohh mentioned, the commands were of String type and some additional manipulation was required before using them in a switch statement. This seems to be more verbose and tedious as compared to simply using `static final` constants.

I can't help but wonder if there's any benefit to using enums in this scenario instead of constants. Any thoughts? 🤔

**Comment by JonahhGohh** [comment]
@chesterhow I agree, it does seem quite verbose to have additional manipulation to support the enum type. But I think some possible benefits enum could have over stacking `static final` constants in main is that it provides better readability, and added functionality (enum methods).

**Comment by samuelfangjw** [comment]
I used enums for the commands as well, but also had to do some manipulation to get the enum constant. I converted the input string to uppercase and used the `Enum.valueOf()` method to get the corresponding enum constant, catching the IllegalArgumentException that is thrown if the command is not valid. This had the added advantage of allowing the commands to be case insensitive.

**Comment by figo2127** [comment]
I agree with Samuel. And your code will look much more readable when you implement switch case as you are determining the command eg. todo, deadline etc.

**Comment by rajobasu** [comment]
I have mostly used enums with a string property in them, which allows for easy for-each style iteration while determining the type of Event to be generated.

**Comment by Assyarul** [comment]
Initially, I thought of using enums for easy iteration as well, but it still presents the problem of poor readability ( I really don't like the long switch/if-else statements presented as we added more and more keywords ). I opted for using a HashMap instead and eventually led to using the [Command Pattern](https://nus-cs2103-ay2021s2.github.io/website/se-book-adapted/chapters/designPatterns.html#command-pattern) presented in the book.

**Comment by nighoggDatatype** [comment]
tood, deadline, event have all very similar code, so we could probably use enum for that.

**Comment by nighoggDatatype** [comment]
I sincerely think that trying to use enums is a poor idea. Using exceptions is useful as a pseudo goto instructions, but I think enums are too clunky to implement right now. 

Also, am I right to say that we can skip doing enums and still count as a completed project?

**Comment by tjtanjin** [comment]
> Also, am I right to say that we can skip doing enums and still count as a completed project?

Prof Damith mentioned in a separate issue [here](https://github.com/nus-cs2103-AY2021S2/forum/issues/12) that there are no penalties for `optional` and `if-applicable` increments :)

**Comment by marcusleeeugene** [comment]
I used enum class for different exceptions types: E.g. INVALID_INTEGER, EMPTY_DESCRIPTION and passed them as parameters into my exception class constructors when throwing exceptions

**Comment by rajobasu** [comment]
@marcusleeeugene using enums for exceptions are probably a bad choice (?!), instead those should be ExceptionClasses themselves, like InvalidIntegerException, or EmptyDescriptionException, if you so want it to be!

**Comment by w2vgd** [comment]
I do not think that using enums for exceptions is wrong by any means, but I do agree that using enums for exceptions does not seem to be the best choice. An enum cannot extend any other class, and in particular, it cannot extend the Exception class (or Throwable class), which means that it will not be able to inherit any methods/fields of the Exception/Throwable classes (and also you won't be able to use the `throw` keyword too). You will have to implement them yourself if you choose to create a standalone enum.
